src/tests.ts

```diff
@@ -1,13 +1,57 @@
 import * as assert from 'assert';
+import { pipe, pipeWith } from 'pipe-ts';
+import * as urlHelpers from 'url';
 import {
     addQueryToUrl,
     appendPathnameToUrl,
+    mapParsedUrl,
+    mapUrl,
     replaceHashInUrl,
     replacePathInUrl,
+    replacePathnameInParsedUrl,
     replacePathnameInUrl,
+    replaceQueryInParsedUrl,
     replaceQueryInUrl,
 } from './index';
 
+assert.deepEqual(
+    pipeWith(
+        urlHelpers.parse('https://foo.com/bar', true),
+        parsedUrl =>
+            mapParsedUrl(({ parsedUrl }) => ({
+                ...parsedUrl,
+                pathname: '/foo',
+                query: { a: 'b' },
+            }))({ parsedUrl }),
+        urlHelpers.format,
+    ),
+    'https://foo.com/foo?a=b',
+);
+
+assert.deepEqual(
+    pipeWith('https://foo.com/bar', url =>
+        mapUrl(({ parsedUrl }) => ({
+            ...parsedUrl,
+            pathname: '/foo',
+            query: { a: 'b' },
+        }))({ url }),
+    ),
+    'https://foo.com/foo?a=b',
+);
+
+assert.deepEqual(
+    pipeWith('https://foo.com/bar', url =>
+        mapUrl(
+            pipe(
+                replacePathnameInParsedUrl({ newPathname: () => '/foo' }),
+                parsedUrl =>
+                    replaceQueryInParsedUrl({ newQuery: () => ({ a: 'b' }) })({ parsedUrl }),
+            ),
+        )({ url }),
+    ),
+    'https://foo.com/foo?a=b',
```

Comment:
When running multiple operations on a URL all at once (via `mapUrl`), you can also compose update functions (a bit like transducers).